### PR TITLE
96 correct object position using height

### PR DESCRIPTION
**Pull request by Greve2001**

### CODE REVIEW

File: remote/src/main/java/courseObjects/Course.java [hunk 1]
```diff
@@ -10,80 +10,71 @@
...
+    public Course(double cameraHeight){
+        this.cameraHeight = cameraHeight;
     }
 
-
     // Getters and setters
-    public synchronized Point getTopLeft() {
-        return topLeft;
+    public Border getBorder() {
+        return border;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:30]:
```suggestion
        return this.border;
```

File: remote/src/main/java/courseObjects/Course.java [hunk 2]
```diff
@@ -10,80 +10,71 @@
...
     }
 
-    public synchronized Point getTopRight() {
-        return topRight;
-    }
-
-    public synchronized void setTopRight(Point topRight) {
-        this.topRight = topRight;
+    public synchronized List<Ball> getBalls() {
+        return balls;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:38]:
```suggestion
        return this.balls;
```

File: remote/src/main/java/courseObjects/Course.java [hunk 3]
```diff
@@ -10,80 +10,71 @@
...
-    public synchronized Point getBottomLeft() {
-        return bottomLeft;
+    public synchronized void setBalls(List<Ball> balls) {
+        this.balls = Collections.synchronizedList(balls);
     }
 
-    public synchronized void setBottomLeft(Point bottomLeft) {
-        this.bottomLeft = bottomLeft;
+    public synchronized Robot getRobot() {
+        return robot;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:46]:
```suggestion
        return this.robot;
```

File: remote/src/main/java/courseObjects/Course.java [hunk 4]
```diff
@@ -10,80 +10,71 @@
...
-    public synchronized Point getBottomRight() {
-        return bottomRight;
+    public synchronized void setRobot(Robot robot) {
+        this.robot = robot;
     }
 
-    public synchronized void setBottomRight(Point bottomRight) {
-        this.bottomRight = bottomRight;
+    public double getCameraHeight() {
+        return cameraHeight;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:54]:
```suggestion
        return this.cameraHeight;
```

File: remote/src/main/java/courseObjects/Course.java [hunk 5]
```diff
@@ -10,80 +10,71 @@
...
     }
 
-    public synchronized List<Ball> getBalls() {
-        return balls;
+    public Point getCameraPosition() {
+        return cameraPosition;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:58]:
```suggestion
        return this.cameraPosition;
```

File: remote/src/main/java/courseObjects/Course.java [hunk 6]
```diff
@@ -10,80 +10,71 @@
  * (3) a list containing the balls located on the course to be picked up by the robot, and (4) the robot object
  */
 public class Course {
-    private final double length, width; // Measured from the innermost sides
-    private Point topLeft, topRight, bottomLeft, bottomRight; // The corners of the border.
+    private final double width = 169.0, height = 123.7; // Measured from the innermost sides
+    private Point cameraPosition;
+
+    private final int resolutionWidth = 1024;
+    private final int ResolutionHeight = 768;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:17]: Variables should start with small letter

File: remote/src/main/java/courseObjects/Course.java [hunk 7]
```diff
@@ -10,80 +10,71 @@
...
-    public synchronized void setBalls(List<Ball> balls) {
-        this.balls = Collections.synchronizedList(balls);
+    public void setCameraPosition(Point cameraPosition) {
+        this.cameraPosition = cameraPosition;
     }
 
-    public synchronized Robot getRobot() {
-        return robot;
+    public int getResolutionHeight() {
+        return ResolutionHeight;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:66]:
```suggestion
        return this.resolutionHeight;
```

File: remote/src/main/java/courseObjects/Course.java [hunk 8]
```diff
@@ -10,80 +10,71 @@
...
     }
 
-    public synchronized void setRobot(Robot robot) {
-        this.robot = robot;
+    public int getResolutionWidth() {
+        return resolutionWidth;
```
Comment (yech199) [inline remote/src/main/java/courseObjects/Course.java:70]:
```suggestion
        return this.resolutionWidth;
```

File: remote/src/main/java/vision/Algorithms.java [hunk 9]
```diff
@@ -1,8 +1,10 @@
 package vision;
 
 import courseObjects.Ball;
+import courseObjects.Course;
```
Comment (yech199) [inline remote/src/main/java/vision/Algorithms.java:4]:
```suggestion
```

File: remote/src/main/java/vision/DetectionController.java [hunk 10]
```diff
@@ -1,6 +1,7 @@
 package vision;
 
 import courseObjects.Ball;
+import courseObjects.Border;
 import courseObjects.Course;
 import courseObjects.Cross;
 import courseObjects.Robot;
```
Comment (yech199) [inline remote/src/main/java/vision/DetectionController.java:7]:
```suggestion
import courseObjects.*;
```

File: remote/src/main/java/vision/DetectionController.java [hunk 11]
```diff
@@ -176,54 +174,81 @@ private void detectCourse(VideoCapture capture) {
...
+        course.setBalls(correctedBalls);
+
+        // Robot
+        Robot robot = course.getRobot();
+        Point correctedFront = Algorithms.correctedCoordinatesOfObject(robot.getFront(),courseCenter,
+                robot.height, camHeight);
+        Point correctedCenter = Algorithms.correctedCoordinatesOfObject(robot.getCenter(),courseCenter,
+                robot.height, camHeight);
+        double correctedAngle = angleBetweenTwoPoints(correctedCenter.x, correctedCenter.y,
+                correctedFront.x, correctedFront.y);
```
Comment (yech199) [inline remote/src/main/java/vision/DetectionController.java:232]: Not used

File: remote/src/main/java/vision/detection/BorderDetector.java [hunk 12]
```diff
@@ -112,8 +112,9 @@ private BorderSet getBorderFromFrame(Mat frame) {
         // Get offset
         double offsetX = sortedCorners.get(0).x;
         double offsetY = sortedCorners.get(0).y;
+        cameraOffset = new Point(offsetX, offsetY);
```
Comment (yech199) [inline remote/src/main/java/vision/detection/BorderDetector.java:115]:
```suggestion
        this.cameraOffset = new Point(offsetX, offsetY);
```

File: remote/src/main/java/vision/detection/BorderDetector.java [hunk 13]
```diff
@@ -185,8 +186,11 @@ private void updateCross(List<MatOfPoint2f> endPointList) {
         cross.setMiddle(new Point((firstPoint.x + middlePoint.x) / 2, (firstPoint.y + middlePoint.y) / 2));
     }
 
-    public BorderSet getBorderSet() {
-        return this.borderSet;
+    public Border getBorder() {
+        return this.border;
+    }
+    public Point getCameraOffset() {
+        return this.cameraOffset;
```
Comment (yech199) [inline remote/src/main/java/vision/detection/BorderDetector.java:193]:
```suggestion
    public Border getBorder() {
        return this.border;
    }
    
    public Point getCameraOffset() {
        return this.cameraOffset;
```

File: remote/src/main/java/vision/math/Geometry.java [hunk 14]
```diff
@@ -21,4 +21,24 @@ else if (xDiff >= 0 && yDiff <= 0) // Quadrant 4
 
         return angle;
     }
+
+    /**
+     * @param heightCamera Height from floor to camera
+     * @param heightObject Height from floor to measure point on object
+     * @param distanceToCamera distance from object to camera, in a perpendicular line to the camera.
```
Comment (yech199) [inline remote/src/main/java/vision/math/Geometry.java:28]: Parameters in cm

File: remote/src/main/java/vision/math/Geometry.java [hunk 15]
```diff
@@ -21,4 +21,24 @@ else if (xDiff >= 0 && yDiff <= 0) // Quadrant 4
...
+     * @return the distance in centimeter from the camera to the object.
+     */
+    public static double objectActualPosition(double heightCamera, double heightObject, double distanceToCamera){
+        double largeTriangleHypotenuse = Math.sqrt(Math.pow(heightCamera,2)+Math.pow(distanceToCamera,2));
```
Comment (yech199) [inline remote/src/main/java/vision/math/Geometry.java:32]:
```suggestion
        double largeTriangleHypotenuse = Math.sqrt(Math.pow(heightCamera, 2) + Math.pow(distanceToCamera, 2));
```

File: remote/src/main/java/vision/math/Geometry.java [hunk 16]
```diff
@@ -21,4 +21,24 @@ else if (xDiff >= 0 && yDiff <= 0) // Quadrant 4
...
+        if (largeTriangleHypotenuse == 0) return -1; //to prevent division with 0
+        //calculate the angel opposite of the height using the sinus relation
+        double angelHeight = Math.toDegrees(Math.asin(heightCamera/largeTriangleHypotenuse));
```
Comment (yech199) [inline remote/src/main/java/vision/math/Geometry.java:35]:
```suggestion
        double angelHeight = Math.toDegrees(Math.asin(heightCamera / largeTriangleHypotenuse));
```

File: remote/src/main/java/vision/math/Geometry.java [hunk 17]
```diff
@@ -21,4 +21,24 @@ else if (xDiff >= 0 && yDiff <= 0) // Quadrant 4
...
+        double smallTriangleGroundAngel = 90 - angelHeight;
+        if (Math.sin(smallTriangleGroundAngel) == 0) return -1; //to prevent division with 0
+        //finds the length of the small triangle using the sinus relation
+        double smallTriangleLength =
+                (heightObject*Math.sin(Math.toRadians(smallTriangleGroundAngel)))/Math.sin(Math.toRadians(angelHeight));
```
Comment (yech199) [inline remote/src/main/java/vision/math/Geometry.java:40]:
```suggestion
                (heightObject*Math.sin(Math.toRadians(smallTriangleGroundAngel))) / Math.sin(Math.toRadians(angelHeight));
```

File: remote/src/main/java/vision/math/Geometry.java [hunk 18]
```diff
@@ -21,4 +21,24 @@ else if (xDiff >= 0 && yDiff <= 0) // Quadrant 4
...
+
+        return distanceToCamera-smallTriangleLength;
```
Comment (yech199) [inline remote/src/main/java/vision/math/Geometry.java:42]:
```suggestion
        return distanceToCamera - smallTriangleLength;
```

File: remote/src/test/java/AlgorithmsTest.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Test
+    public void correctedCoordinatesOfObjectTest1(){
+        Point  newPoint = Algorithms.correctedCoordinatesOfObject(
+                new Point(20,20),
+                new Point(150, 150),
+                40,
+                165
+        );
+        System.out.println(newPoint.toString());
```
Comment (yech199) [inline remote/src/test/java/AlgorithmsTest.java:24]: What is being tested? Can it be tested with assertEquals, so we don't need to manually check the printed value

File: remote/src/test/java/AlgorithmsTest.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Test
+    public void correctedCoordinatesOfObjectTest2(){
+        Point  newPoint = Algorithms.correctedCoordinatesOfObject(
+                new Point(200,200),
+                new Point(150, 150),
+                40,
+                165
+        );
+        System.out.println(newPoint.toString());
```
Comment (yech199) [inline remote/src/test/java/AlgorithmsTest.java:35]: What is being tested? Can it be tested with assertEquals, so we don't need to manually check the printed value

**Review by yech199 — CHANGES_REQUESTED** [review]

**Review by Greve2001 — COMMENTED** [review]
I have changed everything that i can make sense of

**Review by yech199 — APPROVED** [review]

**Pull request merged**